### PR TITLE
Implement device doctor tool

### PR DESCRIPTION
**Pull request by keyonghan**
As part of devicelab migration to LUCI, we need to support device health check and management in LUCI.

This PR creates a device management tool `device_doctor` to do pre health check before running task, clean up after task finishes, and restart the device when task fails or device is identified as unhealthy.

Related https://github.com/flutter/flutter/issues/66193

### CODE REVIEW

File: device_doctor/lib/src/utils.dart [hunk 1]
```diff
@@ -0,0 +1,461 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
```
Comment (CaseyHillers) [inline device_doctor/lib/src/utils.dart:1]: This file seems like a catch all. Can it be split into it's various pieces (process helpers, health interfaces, config)

File: device_doctor/tool/build.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
+#!/bin/bash
+# Copyright 2020 The Flutter Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+# Fetches corresponding dart sdk from CIPD for different platforms, builds
+# an executable binary of device_doctor to `build` folder.
+#
+# This currently supports linux, mac and windows.
```
Comment (CaseyHillers) [inline device_doctor/tool/build.sh:9]: Does this bash script support Windows, or the package?

File: device_doctor/test/src/health_test.dart [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+  @override
+  Stream<List<int>> get stderr => Stream.fromIterable([
+        [1, 2, 3]
+      ]);
+
+  @override
+  Stream<List<int>> get stdout => Stream.fromIterable([
+        [1, 2, 3]
+      ]);
+}
```
Comment (CaseyHillers) [inline device_doctor/test/src/health_test.dart:172]: Nice work with the tests! It's great to see your improvements to testing infra tools, and it will pay off with less infra outages.

File: device_doctor/lib/src/adb.dart [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+  static const BATTERY_HEALTH_UNKNOWN = 1;
+  static const BATTERY_HEALTH_GOOD = 2;
+  static const BATTERY_HEALTH_OVERHEAT = 3;
+  static const BATTERY_HEALTH_DEAD = 4;
+  static const BATTERY_HEALTH_OVER_VOLTAGE = 5;
+  static const BATTERY_HEALTH_UNSPECIFIED_FAILURE = 6;
+  static const BATTERY_HEALTH_COLD = 7;
+}
+
+class AndroidDeviceDiscovery implements DeviceDiscovery {
```
Comment (jonahwilliams) [inline device_doctor/lib/src/adb.dart:93]: The flutter command line tool has device discovery already built in. Is it possible for you to re-use it?

File: device_doctor/lib/src/adb.dart [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+  @override
+  Future<void> restartDevice() async {
+    //await eval('idevicediagnostics', <String>['restart']);
+    for (Device device in await discoverDevices()) {
+      await device.restart();
+    }
+  }
+}
+
+class NoOpDeviceDiscovery implements DeviceDiscovery {
```
[the remark this answers is not in the document]
Comment (keyonghan) [reply device_doctor/lib/src/adb.dart:333]: This is no longer needed for LUCI side device management. Removed.

File: device_doctor/lib/src/adb.dart [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+    await exec(config.adbPath, ['shell', command]..addAll(arguments), env: env, canFail: false);
+  }
+
+  /// Executes [command] on `adb shell` and returns its standard output as a [String].
+  Future<String> shellEval(String command, List<String> arguments, {Map<String, String> env}) {
+    return eval(config.adbPath, ['shell', command]..addAll(arguments), env: env, canFail: false);
+  }
+}
+
+class IosDeviceDiscovery implements DeviceDiscovery {
```
[the remark this answers is not in the document]
Comment (keyonghan) [reply device_doctor/lib/src/adb.dart]: Done.

File: device_doctor/bin/main.dart [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+      final Map<String, HealthCheckResult> deviceChecks = await devices.checkDevices();
+      final bool hasHealthyDevices = deviceChecks.values.where((HealthCheckResult r) => r.succeeded).isNotEmpty;
+      if (!hasHealthyDevices) {
+        throw StateError('No healthy $_deviceOS device is available for $hostname');
+      }
+      if (_deviceOS == 'ios') {
+        await closeIosDialog();
+      }
+      break;
+    case 'restart':
```
[the remark this answers is not in the document]
Comment (keyonghan) [reply device_doctor/bin/main.dart]: Updated.

File: device_doctor/config.yaml [hunk 8]
```diff
@@ -0,0 +1,33 @@
+# The device operating system for different bots. Possible values: android, ios.
+# This configuration file tracks the Flutter devicelab hardware status, and 
+# needs to be updated whatever changes from real world.
+flutter-devicelab-linux-1: android
```
[the remark this answers is not in the document]
Comment (keyonghan) [reply device_doctor/config.yaml]: Considered using an argument in the very beginning, but somehow switched to a local config file. After revisiting the swarming configurations, I think we should be able to achieve that.
Updated to receive device os as an argument and remove the config file.

File: device_doctor/lib/src/utils.dart (hunk 1 above)
Comment (keyonghan) [reply device_doctor/lib/src/utils.dart:1]: Re-organized the code. Please take another look.

File: device_doctor/tool/build.sh (hunk 2 above)
Comment (keyonghan) [reply device_doctor/tool/build.sh:9]: Yes.

File: device_doctor/lib/src/adb.dart (hunk 4 above)
Comment (keyonghan) [reply device_doctor/lib/src/adb.dart]: Unfortunately the device management tool will be called before task (recipes), where the flutter tool is not ready yet.

File: device_doctor/bin/main.dart [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        '--$actionFlag Supported actions are healthcheck, recover, and cleanup.\n'
+        '--$deviceOSFlag Supported device OS: android or ios.\n');
+    exit(1);
+  }
+
+  Config.initialize(_deviceOS);
+
+  switch (_action) {
+    case 'healthcheck':
+      final Map<String, HealthCheckResult> deviceChecks = await devices.checkDevices();
```
Comment (godofredoc) [inline device_doctor/bin/main.dart:43]: do we need to pass the device type to check the available devices?

File: device_doctor/bin/main.dart [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+      final bool hasHealthyDevices = deviceChecks.values.where((HealthCheckResult r) => r.succeeded).isNotEmpty;
+      if (!hasHealthyDevices) {
+        throw StateError('No healthy $_deviceOS device is available');
+      }
+      if (_deviceOS == 'ios') {
```
Comment (godofredoc) [inline device_doctor/bin/main.dart:48]: This is a recovery step rather than health check. The health check will only include the validation of whether or not a device is showing a system dialog.

File: device_doctor/bin/main.dart [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      }
+      break;
+    case 'recover':
+      await devices.recoverDevice();
+      break;
+    case 'cleanup':
+      await removeCachedData();
+      if (_deviceOS == 'ios') {
+        await removeXcodeDerivedData();
+      }
```
Comment (godofredoc) [inline device_doctor/bin/main.dart:59]: removedCacheData and removeXcodeDerivedData may be part of recoverDevice.

File: device_doctor/bin/main.dart [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+      break;
+    default:
+      print('ERROR: Supported actions are healthcheck, recover, and cleanup.');
```
Comment (godofredoc) [inline device_doctor/bin/main.dart:62]: logs rather than print? this tools will be run by a daemon like process and the logs should not assume stdout console

File: device_doctor/bin/main.dart [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+      break;
+  }
+}
+
+bool _checkArgs(ArgParser parser, List<String> args) {
```
Comment (godofredoc) [inline device_doctor/bin/main.dart:67]: is there a way to tell the argparser that some flags are mandatory rather than manually checking them?

File: device_doctor/lib/src/android_device.dart [hunk 14]
```diff
@@ -0,0 +1,231 @@
...
+  static const BATTERY_HEALTH_UNKNOWN = 1;
+  static const BATTERY_HEALTH_GOOD = 2;
+  static const BATTERY_HEALTH_OVERHEAT = 3;
+  static const BATTERY_HEALTH_DEAD = 4;
+  static const BATTERY_HEALTH_OVER_VOLTAGE = 5;
+  static const BATTERY_HEALTH_UNSPECIFIED_FAILURE = 6;
+  static const BATTERY_HEALTH_COLD = 7;
+}
+
+class AndroidDeviceDiscovery implements DeviceDiscovery {
```
Comment (godofredoc) [inline device_doctor/lib/src/android_device.dart:29]: Here and below docs for class and methods.

File: device_doctor/lib/src/android_device.dart [hunk 15]
```diff
@@ -0,0 +1,231 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (godofredoc) [inline device_doctor/lib/src/android_device.dart:3]: Let's add documentation about the expected dependencies. Eg. this class depends on adb that needs to be shipped independently of this binary.

File: device_doctor/lib/src/android_device.dart [hunk 16]
```diff
@@ -0,0 +1,231 @@
...
+
+  static AndroidDeviceDiscovery _instance;
+
+  Future<String> deviceListOutput() async {
+    return eval(config.adbPath, <String>['devices', '-l'], canFail: false).timeout(Duration(seconds: 15));
+  }
+
+  Future<List<String>> deviceListOutputWithRetries(int retriesDelayMs) async {
+    int retry = 0;
+    while (true) {
```
Comment (godofredoc) [inline device_doctor/lib/src/android_device.dart:51]: use the retry module rather than this custom retry logic.

File: device_doctor/lib/src/config.dart [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+
+import 'device.dart';
+import 'utils.dart';
+
+Config _config;
+Config get config => _config;
+
+class Config {
```
Comment (godofredoc) [inline device_doctor/lib/src/config.dart:16]: Docs? what is this used for?

File: device_doctor/lib/src/config.dart [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+  Config({
+    @required this.deviceOperatingSystem,
+    @required this.hostType,
+  });
+
+  static void initialize(String deviceOS) {
+    String home = Platform.environment['HOME'] ?? Platform.environment['USERPROFILE'];
```
Comment (godofredoc) [inline device_doctor/lib/src/config.dart:23]: what is USERPROFILE and what we need it for?

File: device_doctor/lib/src/config.dart [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+    if (home == null) throw "Unable to find \$HOME or \$USERPROFILE.";
+
+    DeviceOperatingSystem deviceOperatingSystem;
+    switch (deviceOS) {
+      case 'android':
```
Comment (godofredoc) [inline device_doctor/lib/src/config.dart:28]: here we may need to pass something the bot already knows, e.g the dimension for device_os is called iOS

File: device_doctor/lib/src/config.dart [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+        deviceOperatingSystem = DeviceOperatingSystem.android;
+        break;
+      case 'ios':
+        deviceOperatingSystem = DeviceOperatingSystem.ios;
+        break;
+      default:
+        throw BuildFailedError('Unrecognized device_os value: $deviceOS');
+    }
+
+    HostType hostType = HostType.physical;
```
Comment (godofredoc) [inline device_doctor/lib/src/config.dart:38]: This is not needed, it was introduced only to support vms in the agents but not needed for luci.

File: device_doctor/lib/src/config.dart [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+      deviceOperatingSystem: deviceOperatingSystem,
+      hostType: hostType,
+    );
+  }
+
+  final DeviceOperatingSystem deviceOperatingSystem;
+  final HostType hostType;
+
+  String get adbPath {
+    String androidHome = Platform.environment['ANDROID_HOME'];
```
Comment (godofredoc) [inline device_doctor/lib/src/config.dart:50]: This is assuming a full android sdk installation would that be possible from the bot_config environment?

File: device_doctor/lib/src/device.dart [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+DeviceDiscovery get devices => DeviceDiscovery();
+
+/// Type of host that is configured to test.
+enum HostType { vm, physical }
+
+/// Operating system on the devices that this host is configured to test.
+enum DeviceOperatingSystem { android, ios }
+
+/// Discovers available devices and chooses one to work with.
+abstract class DeviceDiscovery {
```
Comment (godofredoc) [inline device_doctor/lib/src/device.dart:21]: This factory class looks very similar to the checks performed in the config, can we consolidate both on them in a single place?

File: device_doctor/lib/src/device.dart [hunk 23]
```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Comment (godofredoc) [inline device_doctor/lib/src/device.dart:4]: Docs? we may want to document this module as the one holding the interfaces/abstract classes definition.

File: device_doctor/lib/src/health.dart [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+
+/// Unlocks the login keychain on macOS.
+///
+/// Whic is required to
+///   1. Enable Xcode to access the certificate for code signing.
+///   2. Mitigate "Your session has expired" issue. See flutter/flutter#17860.
+Future<Null> unlockKeyChain() async {
+  if (Platform.isMacOS) {
+    await exec(
+        'security', <String>['unlock-keychain', '-p', Platform.environment['FLUTTER_USER_SECRET'], 'login.keychain'],
```
Comment (godofredoc) [inline device_doctor/lib/src/health.dart:113]: rm? this logic is not related with either health or recovery.

File: device_doctor/lib/src/health.dart [hunk 25]
```diff
@@ -0,0 +1,116 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
```
Comment (godofredoc) [inline device_doctor/lib/src/health.dart:1]: should this file renamed as recovery? health are just the check of the status, the actions to bring the device to a given state are the recovery ones.

File: device_doctor/lib/src/ios_device.dart [hunk 26]
```diff
@@ -0,0 +1,85 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Comment (godofredoc) [inline device_doctor/lib/src/ios_device.dart:4]: Docs specifying the dependencies of this class?

File: device_doctor/lib/src/ios_device.dart [hunk 27]
```diff
@@ -0,0 +1,85 @@
...
+  }
+
+  IosDeviceDiscovery._();
+
+  static IosDeviceDiscovery _instance;
+
+  @override
+  Future<List<Device>> discoverDevices({int retriesDelayMs = 10000}) async {
+    // Warm up device list. This is to ensure slow devices are discovered.
+    // https://github.com/flutter/flutter/issues/64753
```
Comment (godofredoc) [inline device_doctor/lib/src/ios_device.dart:25]: rm? this code won't be run on the context of the start of a task anymore.

File: device_doctor/lib/src/ios_device.dart [hunk 28]
```diff
@@ -0,0 +1,85 @@
...
+  @override
+  Future<void> recoverDevice() async {
+    for (Device device in await discoverDevices()) {
+      await device.recover();
+    }
+  }
+}
+
+/// iOS device.
+class IosDevice implements Device {
```
Comment (godofredoc) [inline device_doctor/lib/src/ios_device.dart:49]: Are we redefining the Device interface? do we need an interface? e.g are we expecting multiple implementations of this interface?

File: device_doctor/lib/src/list_processes.dart [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+import 'dart:async';
+import 'dart:io';
+
+import 'package:meta/meta.dart';
+import 'package:path/path.dart' as path;
+
+import 'process_helper.dart';
+import 'utils.dart';
+
+/// Returns a list of Process IDs (PIDs).
```
Comment (godofredoc) [inline device_doctor/lib/src/list_processes.dart:14]: from where host or device?

File: device_doctor/lib/src/list_processes.dart [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+  if (Platform.isMacOS || Platform.isLinux) {
+    return _listFlutterProcessesPosix(flutterDirectory);
+  }
+  if (Platform.isWindows) {
+    return _listFlutterProcessesWindows(flutterDirectory);
+  }
+  throw 'Unsupported platform: ${Platform.operatingSystem}';
+}
+
+// POSIX implementation
```
Comment (godofredoc) [inline device_doctor/lib/src/list_processes.dart:25]: Not sure where we will be using this functionality from the context that the tool will be running separately from a task.

File: device_doctor/lib/src/process_helper.dart [hunk 31]
```diff
@@ -0,0 +1,102 @@
+import 'dart:async';
+import 'dart:convert' show utf8, LineSplitter;
```
Comment (godofredoc) [inline device_doctor/lib/src/process_helper.dart:6]: Docs?

File: device_doctor/lib/src/process_helper.dart [hunk 32]
```diff
@@ -0,0 +1,102 @@
...
+  String toString() {
+    return '''
+  command : $command
+  started : $startTime
+  pid     : ${process.pid}
+'''
+        .trim();
+  }
+}
+Future<Process> startProcess(String executable, List<String> arguments,
```
Comment (godofredoc) [inline device_doctor/lib/src/process_helper.dart:30]: Also not sure how are we going to use this class from the context of this tool.

File: device_doctor/lib/src/utils.dart (hunk 1 above)
Comment (godofredoc) [reply device_doctor/lib/src/utils.dart:1]: A lot of the functionality of this file is not needed in the context of this tool can we do a cleanup?

**Review by godofredoc — COMMENTED** [review]
Is very difficult to provide detailed feedback in a huge CL like this one. I'd recommend to split it into multiple pieces.
